cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +718,218 @@ class self_comparator {
 
 }  // namespace equality
 
+namespace hash {
+
+/**
+ * @brief Computes the hash value of an element in the given column.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing the existence of nulls in the table
+ */
+template <template <typename> class hash_function, typename Nullate>
+class element_hasher {
+ public:
+  __device__ element_hasher(
+    Nullate nulls,
+    uint32_t seed             = DEFAULT_HASH_SEED,
+    hash_value_type null_hash = std::numeric_limits<hash_value_type>::max()) noexcept
+    : _check_nulls(nulls), _seed(seed), _null_hash(null_hash)
+  {
+  }
+
+  template <typename T, CUDF_ENABLE_IF(column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    if (_check_nulls && col.is_null(row_index)) { return _null_hash; }
+    return hash_function<T>{_seed}(col.element<T>(row_index));
+  }
+
+  template <typename T, CUDF_ENABLE_IF(not column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    CUDF_UNREACHABLE("Unsupported type in hash.");
+  }
+
+  uint32_t _seed;
+  hash_value_type _null_hash;
+  Nullate _check_nulls;
+};
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing the existence of nulls in the table
+ */
+template <template <typename> class hash_function, typename Nullate>
+class device_row_hasher {
+  friend class row_hasher;
+
+ public:
+  device_row_hasher() = delete;
+
+  __device__ auto operator()(size_type row_index) const noexcept
+  {
+    // Hash the first column w/ the seed
+    auto const initial_hash =
+      cudf::detail::hash_combine(hash_value_type{0},
+                                 type_dispatcher<dispatch_storage_type>(
+                                   _table.column(0).type(),
+                                   element_hasher_adapter<hash_function>{_check_nulls, _seed},
+                                   _table.column(0),
+                                   row_index));
+
+    auto it = thrust::make_transform_iterator(_table.begin(), [=](auto const& column) {
+      return cudf::type_dispatcher<dispatch_storage_type>(
+        column.type(), element_hasher_adapter<hash_function>{_check_nulls}, column, row_index);
+    });
+
+    // Hash each element and combine all the hash values together
+    // note that this starts at 1 and not 0 since we already hashed the first column
+    return detail::accumulate(
+      it + 1, it + _table.num_columns(), initial_hash, [](auto hash, auto h) {
+        return cudf::detail::hash_combine(hash, h);
+      });
+  }
```

Comment:
Building on my comment here: https://github.com/rapidsai/cudf/pull/10641/files#r851336811

I think you'll be doing some favors for the compiler in helping it to inline and optimize this code. 

Also, CTAD is your friend. 

```suggestion
    auto const hasher = element_hasher_adapter{ element_hasher{...} };

    // Hash the first column w/ the seed
    auto const initial_hash =
      cudf::detail::hash_combine(hash_value_type{0},
                                 type_dispatcher<dispatch_storage_type>(
                                   _table.column(0).type(),
                                   hasher,
                                   _table.column(0),
                                   row_index));

    auto it = thrust::make_transform_iterator(_table.begin(), [=](auto const& column) {
      return cudf::type_dispatcher<dispatch_storage_type>(
        column.type(), hasher, column, row_index);
    });

    // Hash each element and combine all the hash values together
    // note that this starts at 1 and not 0 since we already hashed the first column
    return detail::accumulate(
      it + 1, it + _table.num_columns(), initial_hash, [](auto hash, auto h) {
        return cudf::detail::hash_combine(hash, h);
      });
  }
```

Reply:
Also, when dealing with thread local data, it's best not to take args by reference or pointer. That can lead to stack frame usage if the compiler can't inline a function. E.g.,

```

    auto it = thrust::make_transform_iterator(_table.begin(), [=](auto const& column) {
      return cudf::type_dispatcher<dispatch_storage_type>(
        column.type(), hasher, column, row_index);
    });
```

vs
```
    auto it = thrust::make_transform_iterator(_table.begin(), [=](auto const column) {
      return cudf::type_dispatcher<dispatch_storage_type>(
        column.type(), hasher, column, row_index);
    });
```

Reply:
Why does a reference prevent inlining?

Reply:
It doesn't necessarily _prevent_ inlining. But if the function isn't inlined, then taking a reference means there needs to be an object to take the address of. You can't take the address of a register, so this can end up forcing the compiler to put the object in local/stack memory so it can have it's address taken. Passing it by `const` value lets the compiler know it can put it in constant banks instead of stack/local memory.